app/scripts/controllers/plugins.js

```diff
@@ -0,0 +1,149 @@
+const ObservableStore = require('obs-store')
+const EventEmitter = require('safe-event-emitter')
+const extend = require('xtend')
+const SES = require('ses')
+const Box = require('3box/dist/3box.min')
+
+class PluginsController extends EventEmitter {
+
+  constructor (opts = {}) {
+    super()
+    const initState = extend({
+      plugins: {},
+      pluginStates: {},
+    }, opts.initState)
+    this.store = new ObservableStore(initState)
+
+    this.provider = opts.provider
+    console.log('!!!!! opts', opts)
+    this._onUnlock = opts._onUnlock
+    this._onNewTx = opts._onNewTx
+    this._subscribeToPreferencesControllerChanges = opts._subscribeToPreferencesControllerChanges
+    this._updatePreferencesControllerState = opts._updatePreferencesControllerState
+    this._signPersonalMessage = opts._signPersonalMessage
+    this._getAccounts = opts._getAccounts
+  }
+
+  get (pluginName) {
+    return this.store.getState().plugins[pluginName]
+  }
+
+  // When a plugin is first created, where should it be executed?
+    // And how do we ensure that the same plugin is never executed twice?
+
+  updatePluginState (pluginName, newPluginState) {
+    const plugins = this.store.getState().plugins
+    const plugin = plugins[pluginName]
+    const updatedPlugin = { ...plugin, pluginState: newPluginState }
+
+    const newPlugins = {...plugins, [pluginName]: updatedPlugin}
+
+    this.store.updateState({
+      plugins: newPlugins,
+    })
+  }
+
+  async create (pluginName) {
+    const plugins = this.store.getState().plugins
+
+    // if (plugins[pluginName]) {
+    //   return plugins[pluginName]
+    // }
+
+    const { source, uiWrappers, requestedAPIs } = await this._getPluginConfig(pluginName)
+    const accounts = await this._getAccounts()
+    console.log('!!!!!* p accounts', accounts)
+    const address = accounts[0]
+
+    const newPlugin = {
+      handleRpcRequest: async (result) => {
+        return Promise.resolve(result)
+      },
+      pluginName,
+      source,
+      pluginState: {},
+      uiWrappers,
+    }
+
+    const newPlugins = {...plugins, [pluginName]: newPlugin}
+
+    this.store.updateState({
+      plugins: newPlugins,
+    })
+
+    this._startPlugin(pluginName, source, requestedAPIs, address)
+
+    return newPlugin
+  }
+
+  _generateApisToProvide (apiList, pluginName) {
+    const updatePluginState = this.updatePluginState.bind(this, pluginName)
+    const possibleApis = {
+      fetch,
+      updatePluginState,
+      onNewTx: this._onNewTx,
+      onUnlock: this._onUnlock,
+      Box,
+      subscribeToPreferencesControllerChanges: this._subscribeToPreferencesControllerChanges,
+      updatePreferencesControllerState: this._updatePreferencesControllerState,
+      generateSignature: (text, address) => {
+        const msg = '0x' + Buffer.from(text, 'utf8').toString('hex')
+
+        return this._signPersonalMessage({
```

Comment:
Why not expose `personal_sign` directly? Just for hex prefixing? If that's what, we should just add that to our personal_sign pre-processing :)

Reply:
Is that needed in all instances of personal_sign? If not, this could also be moved to the plugin code. You are right that we can just provide it directly